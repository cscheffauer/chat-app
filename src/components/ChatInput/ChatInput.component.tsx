import React, { FormEvent, KeyboardEvent } from 'react';

import './ChatInput.scss';
import { AppContext } from '../../AppContextProvider';

interface Props {
	parentProps: {
		messageRef: React.RefObject<HTMLInputElement>;
		message: string;
		setmessage: any;
		editmode: boolean;
		messageidtoedit: string;
		originalmessage: string;
		cancelEditMode: () => void;
	};
}

const ChatInput = ({ parentProps }: Props) => {
	const { send, userid, username } = React.useContext(AppContext);
	const { editmode, messageidtoedit, originalmessage, cancelEditMode } = parentProps;
	const { messageRef, message, setmessage } = parentProps;

	const handleChange = (event: FormEvent<HTMLInputElement>) => {
		event.preventDefault();
		setmessage(event.currentTarget.value);
	};

	const handleSubmit = (event: FormEvent | KeyboardEvent) => {
		event.preventDefault();
		if (message.length > 0) {
			if (editmode) {
				if (originalmessage !== message) {
					send({ id: messageidtoedit, text: message, type: 'editmessageevent' }); //send editmessageevent to backend
					cancelEditMode();
				}
			} else {
				const sent = new Date();
				const messageJson = { message, username, userid, sent };
				send({ message: messageJson, type: 'newmessageevent' }); //send newmessageevent to backend
				setmessage('');
			}
		}
	};

	return (
		<div className='chatInput'>
			<form onSubmit={handleSubmit}>
				<input ref={messageRef} placeholder={'Message'} onChange={handleChange} onBlur={cancelEditMode} value={message} type='text'></input>
				{editmode && <span>âœŽ</span>}
			</form>
		</div>
	);
};

export default ChatInput;
