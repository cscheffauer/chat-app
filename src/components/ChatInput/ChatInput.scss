@import '../../variables';

.chatInput {
	height: 8vh; /* Fallback for browsers that do not support Custom Properties */
	height: calc(var(--vh, 1vh) * 8);
	display: flex;
	align-items: center;
	justify-content: center;
	form {
		width: 94%;
		position: relative;
		input {
			border-radius: 5px;
			border: 1px solid $color-light-grey;
			font-size: 1.2em;
			height: 5vh; /* Fallback for browsers that do not support Custom Properties */
			height: calc(var(--vh, 1vh) * 5);
			width: 100%;
			box-sizing: border-box;
			padding-left: 12px;
			padding-right: 35px;
		}
		span {
			position: absolute;
			font-size: 18px;
			color: $color-grey;
			top: calc(50% - 11px);
			right: 11px;
		}
	}
}
