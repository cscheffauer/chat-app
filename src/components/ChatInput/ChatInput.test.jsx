import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import ChatInput from './ChatInput.component';


// TODO: rework test with React Context https://stackoverflow.com/questions/54292298/jest-mock-react-context
describe('<ChatInput />', () => {
    
	let wrapper;
	const useridMock = '1234';
	const messagesMock = [
		{ username: 'Max', text: 'Hallo', sent: new Date('2020-01-01'), userid: '1', messageid: '1', state: '' },
		{ username: 'John', text: 'Hi', sent: new Date('2020-01-01'), userid: '2', messageid: '2', state: 'DELETED' },
		{ username: 'Jeffrey', text: "What's up?", sent: new Date('2020-01-01'), userid: '3', messageid: '3', state: 'EDITED' },
	];

	const sendMessageMock = jest.fn();
	const editMessage = jest.fn();
	const deleteMessage = jest.fn();

	beforeEach(() => {
		wrapper = mount(<ChatInput sendMessage={sendMessageMock} editMessage={editMessage} deleteMessage={deleteMessage} userid={useridMock} messages={messagesMock} />);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('expect to render ChatInput component', () => {
		expect(wrapper).toMatchSnapshot();
	});
	describe('Testing the Message Functions', () => {
		it('should send the message value on submit', () => {
			const chatInput = wrapper.find('input');
			chatInput.instance().value = 'Hello everyone!';
			chatInput.simulate('change');
			act(() => {
				wrapper
					.find('form')
					.props()
					.onSubmit({ preventDefault() {} }); //submit form
			});

			expect(sendMessageMock).toHaveBeenCalledWith('Hello everyone!');
		});
	});
});