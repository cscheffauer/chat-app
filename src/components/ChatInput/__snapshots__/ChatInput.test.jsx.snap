// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatInput /> expect to render Chat component 1`] = `undefined`;
